File: PrimePairWithTargetSum/PrimePairWithTargetSum.cpp
/*
https://www.geeksforgeeks.org/problems/sum-of-prime4751/1?utm_source=geeksforgeeks&utm_medium=newui_home&utm_campaign=potd
*/

//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

class Array {
  public:
    template <class T>
    static void input(vector<T> &A, int n) {
        for (int i = 0; i < n; i++) {
            scanf("%d ", &A[i]);
        }
    }

    template <class T>
    static void print(vector<T> &A) {
        for (int i = 0; i < A.size(); i++) {
            cout << A[i] << " ";
        }
        cout << endl;
    }
};


// } Driver Code Ends

class Solution {
  public:
  
    void sieve(vector<int>&allPrimes, int n){
        vector<int>isPrime(n+10, true);
        isPrime[0]=false;
        isPrime[1]=false;
        for(int i=2; i*i<=n; ++i){
            if(isPrime[i]){
                for(int j=2; i*j<=n; ++j){
                    isPrime[i*j]=false;
                }
            }
        }
        for(int i=1; i<=n; ++i){
            if(isPrime[i]){
                allPrimes.push_back(i);
            }
        }
    }
  
    vector<int> getPrimes(int n) {
        vector<int>allPrimes;
        sieve(allPrimes, n);
        int left=0, right=allPrimes.size()-1;
        while(left<=right){
            int cursm=allPrimes[left]+allPrimes[right];
            if(cursm==n){
                return {allPrimes[left], allPrimes[right]};
            } else if(cursm>n){
                right-=1;
            } else left+=1;
        }
        return {-1, -1};
    }
};


//{ Driver Code Starts.

int main() {
    int t;
    scanf("%d ", &t);
    while (t--) {

        int n;
        scanf("%d", &n);

        Solution obj;
        vector<int> res = obj.getPrimes(n);

        Array::print(res);
    }
}

// } Driver Code Ends

File: PrimePairWithTargetSum/PrimePairWithTargetSum.java
/*
https://www.geeksforgeeks.org/problems/sum-of-prime4751/1?utm_source=geeksforgeeks&utm_medium=newui_home&utm_campaign=potd
*/

//{ Driver Code Starts
import java.io.*;
import java.util.*;
import java.util.ArrayList;

class IntArray {
    public static int[] input(BufferedReader br, int n) throws IOException {
        String[] s = br.readLine().trim().split(" ");
        int[] a = new int[n];
        for (int i = 0; i < n; i++) a[i] = Integer.parseInt(s[i]);

        return a;
    }

    public static void print(int[] a) {
        for (int e : a) System.out.print(e + " ");
        System.out.println();
    }

    public static void print(ArrayList<Integer> a) {
        for (int e : a) System.out.print(e + " ");
        System.out.println();
    }
}

class GFG {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t;
        t = Integer.parseInt(br.readLine());
        while (t-- > 0) {

            int n;
            n = Integer.parseInt(br.readLine());

            Solution obj = new Solution();
            ArrayList<Integer> res = obj.getPrimes(n);

            IntArray.print(res);
        }
    }
}

// } Driver Code Ends


class Solution {
    
    public static void sieveOfEratosthenes(int N, ArrayList<Integer>allPrimesUntilN, ArrayList<Boolean>isTheNumberPrime){
        for(int i=2; i*i<=N; ++i){
            if(isTheNumberPrime.get(i)==true){
                for(int j=2; i*j<=N; ++j){
                    isTheNumberPrime.set(i*j, false);
                }
            }
        }
        for(int i=1; i<=N; ++i){
            if(isTheNumberPrime.get(i)==true) allPrimesUntilN.add(i);
        }
    }
    
    public static void generateAllPrimesUntilN(int N, ArrayList<Integer>allPrimesUntilN){
        ArrayList<Boolean>isTheNumberPrime=new ArrayList<>();
        for(int i=0; i<=N+10; ++i){
            isTheNumberPrime.add(true);
        }
        isTheNumberPrime.set(0, false);
        isTheNumberPrime.set(1, false);
        sieveOfEratosthenes(N, allPrimesUntilN, isTheNumberPrime);
    }
    
    public static ArrayList<Integer> getPrimes(int n) {
        ArrayList<Integer>allPrimesUntilN=new ArrayList<>();
        ArrayList<Integer>resultantPrimePair=new ArrayList<>();
        generateAllPrimesUntilN(n, allPrimesUntilN);
        int leftIndex=0, rightIndex=allPrimesUntilN.size()-1;
        while(leftIndex<=rightIndex){
            int sumOfPrimesAtCurrentLeftRightIndex=allPrimesUntilN.get(leftIndex)+allPrimesUntilN.get(rightIndex);
            if(sumOfPrimesAtCurrentLeftRightIndex==n){
                resultantPrimePair.add(allPrimesUntilN.get(leftIndex));
                resultantPrimePair.add(allPrimesUntilN.get(rightIndex));
                break;
            } else if(sumOfPrimesAtCurrentLeftRightIndex>n){
                rightIndex-=1;
            } else leftIndex+=1;
        }
        if(resultantPrimePair.size()==0){
            resultantPrimePair.add(-1);
            resultantPrimePair.add(-1);
        }
        return resultantPrimePair;
    }
}
